MdeModulePkg/Core/Dxe/Misc/MemoryProtection.c

```diff
@@ -727,18 +730,25 @@ GetPermissionAttributeForMemoryType (
   // } else {
   //   return 0;
   // }
+  // MU_CHANGE END
+
+  UINT64  Attributes = 0;
 
   // Handle code allocations according to the NX_COMPAT DLL flag. If the flag is
   // set, the image should update the attributes of code type allocates when it's ready to execute them.
   if (!IsEnhancedMemoryProtectionActive ()) {
     return 0;
-  } else if (GetDxeMemoryTypeSettingFromBitfield (MemoryType, gDxeMps.NxProtectionPolicy)) {
-    return EFI_MEMORY_XP;
   }
 
-  return 0;
+  if (GetDxeMemoryTypeSettingFromBitfield (MemoryType, gDxeMps.NxProtectionPolicy)) {
+    Attributes |= EFI_MEMORY_XP;
+  }
 
-  // MU_CHANGE END
+  if ((MemoryType == EfiConventionalMemory) && gDxeMps.MarkFreeMemoryReadOnly && mPageAttributesInitialized) {
+    Attributes |= EFI_MEMORY_RP;

```

Comment:
I was thinking that we would do this in Page.c itself, on AllocatePages we would remove RP, on FreePages we would set RP. This gets called out from a level of AllocatePages, but is this just to check the HOB? Or to ensure it is happening after the GCD sync? My concern of doing it at this level is that, as earlier in the PR shows, we have code that calls around it and any future code that calls around it would have to do the same thing or it will break.

Reply:
AllocateMemoryProtection policy is how all attributes are set outside of image memory. There were some cases where attributes weren't being applied, but all were contained within DxeCore where the internal functions are accessible outside of a protocol or MemoryAllocationLib. I think we should be out of the woods on those missed allocations -- especially with this update.

What you're suggesting can be done in a future update but I prefer to stick with the current pattern for this.